feat: minimum version check screen

sebastianquek opened the pull request:
~~**Blocked by https://github.com/rationally-app/mobile-application/pull/114**~~

After logging in or being redirected from the login page, users will be taken to a `CampaignInitialisationScreen` that retrieves the latest campaign config and checks for the minimum version.

The campaign config has been setup for pr71 on the backend, so can use that as the endpoint.

Here's a diagram of how this section works:
![image](https://user-images.githubusercontent.com/577802/89139478-5908ef80-d571-11ea-8266-1aaa013392cb.png)

Here's how the error screens look:
|![Simulator Screen Shot - iPhone 8 - 2020-08-13 at 14 02 07](https://user-images.githubusercontent.com/577802/90099731-b4a35c00-dd6d-11ea-94d4-9824baae7113.png)|![Simulator Screen Shot - iPhone 8 - 2020-08-13 at 14 01 49](https://user-images.githubusercontent.com/577802/90099737-b705b600-dd6d-11ea-90a9-b8308fc62cad.png)|
|--|--|


## To test the different flows
1. Set the version of the app by changing the passed in variables in `npm start`, refer to `package.json`
2. Update the minimum version on the backend param store

szenius commented:
Hmm tested with PR71, a few questions
1. How do I simulate the different scenarios (got update, no update, etc)?
2. Is there a flow diagram that I can refer to? Maybe can add the link in the PR description?

sebastianquek commented:
@szenius I've updated the description to elaborate more on how to test the different flows and also added a flowchart. Let me know if you're still unsure!

szenius commented:
Was caught up with something else earlier on. Testing this now!

sebastianquek commented:
> Got this on sign in. Is it expected?
> 
> ```
> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in %s.%s, a useEffect cleanup function, 
>     in CampaignInitialisationScreen (created by SceneView)
> * [native code]:null in __expoConsoleLog
> - node_modules/react-native/Libraries/LogBox/LogBox.js:33:4 in console.error
> - node_modules/expo/build/environment/muteWarnings.fx.js:27:4 in error
> - node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:608:6 in warningWithoutStack
> - node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:21972:25 in warnAboutUpdateOnUnmountedFiberInDEV
> - node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:19813:40 in scheduleUpdateOnFiber
> - node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:11880:16 in dispatchAction
> * [native code]:null in dispatchAction
> * src/hooks/useUpdateCampaignConfig/useUpdateCampaignConfig.tsx:31:32 in fetchCampaignConfig
> - node_modules/regenerator-runtime/runtime.js:63:36 in tryCatch
> - node_modules/regenerator-runtime/runtime.js:293:29 in invoke
> - node_modules/regenerator-runtime/runtime.js:63:36 in tryCatch
> - node_modules/regenerator-runtime/runtime.js:154:27 in invoke
> - node_modules/regenerator-runtime/runtime.js:164:18 in PromiseImpl.resolve.then$argument_0
> - node_modules/@storybook/addons/node_modules/core-js/modules/es.promise.js:113:24 in microtask$argument_0
> - node_modules/@storybook/addons/node_modules/core-js/internals/microtask.js:25:10 in flush
> - node_modules/promise/setimmediate/core.js:37:13 in tryCallOne
> - node_modules/promise/setimmediate/core.js:123:24 in setImmediate$argument_0
> - node_modules/react-native/Libraries/Core/Timers/JSTimers.js:135:14 in _callTimer
> - node_modules/react-native/Libraries/Core/Timers/JSTimers.js:183:16 in _callImmediatesPass
> - node_modules/react-native/Libraries/Core/Timers/JSTimers.js:446:30 in callImmediates
> * [native code]:null in callImmediates
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:396:6 in __callImmediates
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:144:6 in __guard$argument_0
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:373:10 in __guard
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:4 in flushedQueue
> * [native code]:null in flushedQueue
> * [native code]:null in invokeCallbackAndReturnFlushedQueue
> ```

Oh no, how did you get here?

szenius commented:
@sebastianquek I just logged in, it appeared on the scan ID page~ 

Anyway so far I haven't trigerred the initialisation screen, still trying different cases 🙈

szenius commented on src/components/CampaignInitialisation/CampaignInitialisationScreen.tsx:
```diff
@@ -0,0 +1,160 @@
+
+const RETRY_UPDATE_TIMES = 3;
+
+export const CampaignInitialisationScreen: FunctionComponent<NavigationProps> = ({
```
Decided to start a thread so it's easier to see the conversation!!

Questions
1. Are my configurations correct?
2. Are my observations correct for the following steps?

Did this:
* Left the npm start as is
* Changed `/rationally-api/pr71/client/policies` to have no identifiers in SSM (is this supposed to have an impact? Not very sure)
* Changed `/rationally-api/pr71/client/features` to have `"minAppBuildVersion":3`

What I saw:
* On submit of OTP, a loading screen
* The scan ID screen with the warning

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in %s.%s, a useEffect cleanup function, 
    in CampaignInitialisationScreen (created by SceneView)
* [native code]:null in __expoConsoleLog
- node_modules/react-native/Libraries/LogBox/LogBox.js:33:4 in console.error
- node_modules/expo/build/environment/muteWarnings.fx.js:27:4 in error
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:608:6 in warningWithoutStack
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:21972:25 in warnAboutUpdateOnUnmountedFiberInDEV
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:19813:40 in scheduleUpdateOnFiber
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:11880:16 in dispatchAction
* [native code]:null in dispatchAction
* src/hooks/useUpdateCampaignConfig/useUpdateCampaignConfig.tsx:31:32 in fetchCampaignConfig
- node_modules/regenerator-runtime/runtime.js:63:36 in tryCatch
- node_modules/regenerator-runtime/runtime.js:293:29 in invoke
- node_modules/regenerator-runtime/runtime.js:63:36 in tryCatch
- node_modules/regenerator-runtime/runtime.js:154:27 in invoke
- node_modules/regenerator-runtime/runtime.js:164:18 in PromiseImpl.resolve.then$argument_0
- node_modules/@storybook/addons/node_modules/core-js/modules/es.promise.js:113:24 in microtask$argument_0
- node_modules/@storybook/addons/node_modules/core-js/internals/microtask.js:25:10 in flush
- node_modules/promise/setimmediate/core.js:37:13 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:24 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:135:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:183:16 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:446:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:396:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:144:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:373:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:4 in flushedQueue
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue
```

sebastianquek replied:
Good catch, the issue happens because of this:
1. Enter `CampaignInitialisationScreen`
2. `updateCampaignConfig` is called
3. latest config is fetched, and since it differs, it's saved into async storage + updated context
4. because of this change, `configHashes` is updated
5. `updateCampaignConfig` reference is updated
6. the useEffect in `CampaignInitialisationScreen` triggers `updateCampaignConfig` again
7. at this time, the user has already been navigated to one of the normal screens, and so one of the set states in `CampaignInitialisationScreen` is performed on an already unmounted `CampaignInitialisationScreen`

What I've done is to make an exception to the exhaustive dependencies for this particular `useEffect` as `updateCampaignConfig` should only be called once when this screen is loaded.

Let me know if you know of a better way!

szenius replied:
Solution sounds good! Tested the same flow and it works now without the warning.

szenius commented on src/components/CampaignInitialisation/CampaignInitialisationScreen.tsx:
```diff
@@ -0,0 +1,162 @@
+  useEffect(() => {
+    updateCampaignConfig();
+    // updating the campaign config should only happen once when this screen is loaded
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
I've also done this in other places of the code. I wonder what is the value of this rule in our codebase, since it's not always the best solution 🤔 Just a thought~

sebastianquek replied:
Yea i've avoided this everywhere else, but I think in this case it's a legit one, since we actually really only want this to run once. We can keep this in mind as we develop!

szenius commented on src/components/CampaignInitialisation/CampaignInitialisationScreen.tsx:
```diff
@@ -0,0 +1,162 @@
+
+    // In the case that there's NO_UPDATE, then we probably set the
+    // minimum build version too early, before the app updates have
+    // been released.
```
Hmm a bit confused about what this means. Could you explain when this might happen from a user's perspective? When we say `NO_UPDATE` do we also include `TOO_RECENT_CHECK` since I see that that can also be returned from `useCheckUpdates`?

sebastianquek replied:
This will happen in the case when the user has spotty internet and/or offline, so the latest build couldn't be downloaded after the 3 tries.

`useCheckUpdates` won't return `TOO_RECENT_CHECK` in this case because we pass in the `forceCheck` boolean to `checkUpdates`, so that will check for updates regardless of when the last check was.

Hmmm but this raises the issue of what happens if the fetching of the campaign config didn't succeed in the first place. Should we just show the Error boundary? Should we also try a few times? Should there be some exponential backoff for the retries? 🤔 

What do you think?

szenius replied:
Ohh I understand now. When we are in this part of the code, the premise is that the build is outdated. And yet we see `NO_UPDATE` so that's wrong. Makes sense! 

As for showing error boundary -- if the user's internet is not good in the first place, we are throwing an error from `useCheckVersion` right? Not very familiar with this part of the code but would that already trigger the error boundary?

Anyways I think it might be a good idea to show error boundary if the issue is internet connection, cuz they can't use our app without having internet connection anyway 😆

sebastianquek replied:
Yea I overlooked the error handling before `useCheckVersion`, so when `updateCampaignConfig`, I didn't handle the error.

Good point about the usage of the rest of the app lol, will handle it by showing the error boundary then!

szenius commented on src/components/CampaignInitialisation/CampaignInitialisationScreen.tsx:
```diff
@@ -0,0 +1,162 @@
+        message: `attempt ${i + 1}`
+      });
+      const updateResult = await checkUpdates(true);
+      if (updateResult === "UPDATE_READY") {
```
Would it be worth putting these constants in an object or enum? So we can refer to the same constant from both here and `useCheckUpdates`?

sebastianquek replied:
Hmm this will throw a TypeScript error if "UPDATE_READY" is not within the union type "CheckUpdatesResult". So the union type restricts the values that `updateResult` can take—similar to a typical enum, but without the overhead of needing to do
```
enum CheckUpdatesResult {
  UPDATE_READY = "UPDATE_READY"
  ...
}
vs
type CheckUpdatesResult =
  | "TOO_RECENT_CHECK"
  | "NO_UPDATE"
  | "UPDATE_READY"
  | "UPDATE_ERROR";
```

szenius replied:
Didn't notice the type!! Nice nice

szenius commented on src/components/CampaignInitialisation/CampaignInitialisationScreen.tsx:
```diff
@@ -0,0 +1,162 @@
+    try {
+      const versionCheckResult = checkVersion();
+      switch (versionCheckResult) {
+        case "OK":
```
Same thing about constants here~

sebastianquek replied:
Similar approach to the type here to!

szenius commented on src/components/CampaignInitialisation/CampaignInitialisationScreen.tsx:
```diff
@@ -0,0 +1,162 @@
+  useEffect(() => {
+    if (fetchingState === "RETURNED_NEW_UPDATES") {
+      handleNewCampaignConfig();
+    } else if (fetchingState === "RETURNED_NO_UPDATES") {
```
Same thing about constants here~

szenius replied:
I assume this too? For the type?

szenius commented on src/components/CampaignInitialisation/UpdateFromAppStoreContent.tsx:
```diff
@@ -0,0 +1,63 @@
+
+    case "android":
+    default:
+      // figure out a better way to handle default
```
If this is a TODO, might be useful to add the prefix so we can find this easily~

sebastianquek replied:
Good catch!!

szenius commented on src/components/Login/LoginContainer.tsx:
```diff
@@ -129,10 +129,12 @@ export const InitialisationContainer: FunctionComponent<NavigationProps> = ({
-          break;
         case "MERCHANT":
-          navigation.navigate("MerchantPayoutScreen");
+          // TODO: Navigate directly to CampaignInitialisationScreen on successful login.
```
Seems like a significant refactoring. Curious to know -- what is the impact of this workaround for now?

sebastianquek replied:
Not a lot of impact in terms of functionality. Just that right now the Login component is knowing a lot of things, including details on the features + products. It actually just needs to know if the said user is authenticated and authorised to continue. The retrieval of the campaign config can then be deferred to the CampaignInitialisation component.

By splitting these concerns, in future when users are tagged to multiple campaigns, then we can call the CampaignInitialisation component only when the user selects the campaign. Login will have no part to play in the campaign config.

szenius replied:
Makes sense! I wonder if it is worth making this a chore rather than leave it as a TODO, because I am wary about TODOs just living in the codebase for years after they are added 🙈  But will leave this as a discussion on Thurs~

sebastianquek replied:
Yea, I think for this part, I'll surely move it out when we migrate the backend over to the new campaign config. 🤣
